devito/grid.py

```diff
@@ -0,0 +1,50 @@
+from devito.tools import as_tuple
+from devito.dimension import x, y, z
+
+import numpy as np
+
+__all__ = ['Grid']
+
+
+class Grid(object):
+
+    """
+    A cartesian grid that encapsulates a physical domain over which
+    to discretize :class:`Function`s.
+
+    :param shape: Shape of the domain region in grid points.
+    :param extent: Physical extent of the domain in m; defaults to a
+                   unit box of extent 1m in all dimensions.
+    :param origin: Physical coordinate of the origin of the domain;
+                   defaults to 0. in all dimensions.
+    :param dtype: Default data type to be inherited by all Functions
+                  created from this :class:`Grid`.
+    """
```

Comment:
Convention is usually (n,d,o)
n : number of grid points
d : spacing
o : origin

the get extent from this three. The current way will give wrong spacing if `origin` is non-zero.

Reply:
Yeah, so I don't think the current convention is particularly intuitive, tbh. My primary concern is that the `Grid` primarily defines a physical domain, given by the origin (bottom left in 2D) and the extent (or "width") of the domain, so that the top right corner is defined as `origin + extent`. I'll make this clearer in the docstring.

The other thing I would really like is that the `spacing` is a diagnostic property determined by the spatial extent of the domain and the number of either cells or nodes (currently nodes) that define the grid shape. This effectively makes it much easier to create equivalent physical grids with different discretisations without the need for the user to carefully check that `spacing * shape` always comes out to the same exact `extent`. After all, a unit-cube always has `extent=(1., 1., 1.)`, no matter what `shape` I discretise it with - and I believe this mirrors general PDE-conventions more closely.